lib/ride/services/audio.dart

```diff
@@ -0,0 +1,820 @@
+    }
+  }
+
+  /// Add the current speed value to the last speed values list.
+  _addSpeedValueToLastSpeedValues() {
+    if (positioning!.lastPosition == null) {
+      return;
+    }
+    if (positioning!.lastPosition!.speed < 5) {
+      return;
+    }
+
+    if (lastSpeedValues.length >= 11) {
+      lastSpeedValues.removeAt(0);
+    }
+    lastSpeedValues.add(positioning!.lastPosition?.speed ?? 0);
+  }
+
+  /// Returns the median speed of the last speed values.
+  double _getMedianSpeedOfLastSpeedValues() {
+    if (lastSpeedValues.isEmpty) {
+      return 0;
+    }
+    lastSpeedValues.sort();
+    if (lastSpeedValues.length % 2 == 0) {
+      return (lastSpeedValues[lastSpeedValues.length ~/ 2 - 1] + lastSpeedValues[lastSpeedValues.length ~/ 2]) / 2;
+    } else {
+      return lastSpeedValues[lastSpeedValues.length ~/ 2];
+    }
+  }
+
+  /// Returns a list with the change times of the recommendation.
+  List<int> _calculateChangesTimesOfRecommendation(Recommendation recommendation) {
+    List<int> changeTimes = [];
+    Phase lastPhase = recommendation.calcCurrentSignalPhase;
+    for (int i = 0; i < recommendation.calcPhasesFromNow.length; i++) {
+      // Only consider red and green.
+      if (!(recommendation.calcPhasesFromNow[i] == Phase.red || recommendation.calcPhasesFromNow[i] == Phase.green)) {
+        continue;
+      }
+
+      if (recommendation.calcPhasesFromNow[i] != lastPhase) {
+        changeTimes.add(i);
+        lastPhase = recommendation.calcPhasesFromNow[i];
+      }
+    }
+    return changeTimes;
+  }
+
+  /// Returns the closest change time in reach.
+  int _getClosestChangeTimeInReach(
+      int arrivalTime, List<int> changeTimes, SpeedMode speedMode, double distanceToNextSg) {
+    double maxSpeed = (speedMode.maxSpeed - 3) / 3.6; // Convert to m/s.
+
+    return changeTimes
+        .reduce((a, b) => (a - arrivalTime).abs() < (b - arrivalTime).abs() && distanceToNextSg / a < maxSpeed ? a : b);
+  }
+
+  /// Check if instruction contains sg information and if so add countdown.
+  /// Speed in m/s.
+  InstructionText? generateTextToPlay(InstructionText instructionText) {
+    ride ??= getIt<Ride>();
+
+    // Check if prediction quality is not good enough.
+    if (ride!.calcCurrentSG == null ||
+        ride!.predictionProvider?.recommendation == null ||
+        (ride!.predictionProvider?.prediction?.predictionQuality ?? 0) < predictionQualityThreshold) {
+      // No sg countdown information can be added and thus instruction part must not be played.
+      return null;
+    }
+
+    final recommendation = ride!.predictionProvider!.recommendation!;
+
+    List<int> changeTimes = _calculateChangesTimesOfRecommendation(recommendation);
+
+    if (changeTimes.isEmpty) return null;
+
+    double lastMedianSpeed = _getMedianSpeedOfLastSpeedValues();
+
+    // Check if the median speed is less then 1.5m/s and replace with 5m/s.
+    if (lastMedianSpeed < 1.5) lastMedianSpeed = 5;
```

Comment:
This should be explained. For someone who didn't develop this (me), this makes no sense.

Reply:
Also, why use the median of the last 10 speeds? Isn't the exact last one the most relevant?